### ISSUE TITLE
COST() ignores inferred_tolerance_default

### ISSUE THREAD

**Issue by blais** [beancount/beancount#361]
**[Original report](https://bitbucket.org/blais/beancount/issue/361) by Martin Michlmayr (Bitbucket: [tbm13](https://bitbucket.org/tbm13), GitHub: [tbm](https://github.com/tbm)).**

----------------------------------------

Let's take this example

```
2000-01-01 open Liabilities:Accounts-Payable
2000-01-01 open Expenses:Travel

2013-05-17 * "Test" "Train 1"
    Expenses:Travel
    Liabilities:Accounts-Payable     -2.85 EUR {1.2817 USD, 2013-05-17}

2013-06-17 * "Test" "Train 2"
    Expenses:Travel
    Liabilities:Accounts-Payable     -2.95 EUR {1.283 USD, 2013-05-17}
```

Now:

```
beancount> select date, position, balance where account ~ 'Expenses:Travel'
   date     position   balance  
---------- ---------- ----------
2013-05-17 3.6528 USD 3.6528 USD
2013-06-17 3.7848 USD 7.4376 USD
beancount> 
```

So the precision is quite high.  Ok, let's add

```
option "inferred_tolerance_default" "USD:0.01"
```

and now we get:

```
beancount> select date, position, balance where account ~ 'Expenses:Travel'
   date     position   balance  
---------- ---------- ----------
2013-05-17 3.65   USD 3.65   USD
2013-06-17 3.78   USD 7.43   USD
```

Ok, good.

This is for Expenses:.  If we look at Liabilities, we should get the same result.

But wait:

```
beancount> select date, COST(position) where account ~ 'Liabilities:Accounts-Payable'
   date    cost_positi
---------- -----------
2013-05-17 -3.6528 USD
2013-06-17 -3.7848 USD
```

We still have all that precision.

And worse:

```

beancount> select SUM(COST(position)) where account ~ 'Liabilities:Accounts-Payable'
sum_cost_po
-----------
-7.4376 USD
```

which after rounding is 7.44.  But expenses were 7.43!

**Comment by blais** [comment]
**Original comment by Martin Michlmayr (Bitbucket: [tbm13](https://bitbucket.org/tbm13), GitHub: [tbm](https://github.com/tbm)).**

----------------------------------------

Related to issue #235

**Comment by blais** [comment]
**Original comment by Martin Michlmayr (Bitbucket: [tbm13](https://bitbucket.org/tbm13), GitHub: [tbm](https://github.com/tbm)).**

----------------------------------------

If we add:

```
plugin "beancount.plugins.implicit_prices"
```

we get:

```
beancount> select CONVERT(position, 'USD', date) where account ~ 'Liabilities:Accounts-Payable'
convert_pos
-----------
-3.6528 USD
-3.7848 USD
beancount> select SUM(CONVERT(position, 'USD', date)) where account ~ 'Liabilities:Accounts-Payable'
sum_convert
-----------
-7.4376 USD
```

I'd also aspect rounding here

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

I think you may be misunderstanding what the tolerances does.
First, please note the important difference between tolerance and precision. Tolerance is used to decide how to round up/down missing amounts.
Precision is used to round numbers /for rendering only'.

Case in point, in the beginning of your question, let's examine what your option does. The first transaction is parsed as this (I'm using bean-doctor context):

------------ Transaction


```
#!python

2013-05-17 * "Test" "Train 1"
  Expenses:Travel               3.6528 USD                           ;  3.652845 USD
  Liabilities:Accounts-Payable   -2.85 EUR {1.2817 USD, 2013-05-17}  ; -3.652845 USD


Tolerances: EUR=0.005
Basis: (-3.652845 USD)
```


When you add the tolerance option, it becomes this:

------------ Transaction


```
#!python

2013-05-17 * "Test" "Train 1"
  Expenses:Travel               3.6500 USD                           ;      3.65 USD
  Liabilities:Accounts-Payable   -2.85 EUR {1.2817 USD, 2013-05-17}  ; -3.652845 USD


Residual: (-0.002845 USD)
Tolerances: EUR=0.005, USD=0.01
Basis: (-3.652845 USD)

```

As you can see, only the expenses postings change.
When you select the liabilities postings, they still have their full cost.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Not an issue.

**Comment by blais** [comment]
**Original comment by Martin Michlmayr (Bitbucket: [tbm13](https://bitbucket.org/tbm13), GitHub: [tbm](https://github.com/tbm)).**

----------------------------------------

Thanks for explaining the difference between tolerance and precision.

Anyway, I disagree with closing this as "not an issue" since it's clearly an issue. I assume there's currently no way in beancount to force the output I want?

You say:

> When you select the liabilities postings, they still have their full cost

But why?  I know that 2.85 EUR held at 1.2817 USD cost is 3.652845 USD.  But why doesn't my tolerance apply to that?  Ok, you explain it "Tolerance is used to decide how to round up/down missing amounts" -- it's not a missing amount.  I specified the amount.

So clearly we need a way to force the precision (and not just display precision, but actual rounding).  I think you mentioned you are planning to add such an option.

I think this ticket should be kept open until such an option exists. It shows a good use case.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

I'm very confused, I'm not sure which number you're talking about.
2.85 is stored as is.
1.2817 also is.
If you enter a number, I don't round it.
Tolerances are used to decide whether a residual amount from balancing a transaction is an error (below the tolerated amount, it's not an error).
If a number is elided from the input, that tolerance is also used to round the number.
You can use the "bean-query <filename> print" command to inspect the actual effect of these options in detail, on each of the postings.

Are you suggesting that we should round out the total cost somehow?
Some amount of round may occur on rendering, but that's only for display.
(The original cost, when multiplied by the number of units, will never produce a number that has the desired precision of the currency.)

Where's the problem exactly?
How is this an issue?

**Comment by blais** [comment]
**Original comment by Martin Michlmayr (Bitbucket: [tbm13](https://bitbucket.org/tbm13), GitHub: [tbm](https://github.com/tbm)).**

----------------------------------------

Martin, thanks for your elaboration.  I see that we're approaching this from different perspectives.  Let me explain where I'm coming from and give some use cases and maybe we can work something out.

It's true that I specify the cost with high precision.  However, I believe there are good reasons to use higher precision for the cost and to have that rounded later.

Let me give a different example that makes more sense than my original example.  On some stock markets, shares or funds are quoted with a fairly high precision.  Let's say that 1 FOO costs 1.2817 USD.

Now if I actually purchase 1 FOO, I'm not going to pay $1.2817.  I'll pay $1.28, or maybe $1.29.  You would argue that I should use that value for the cost:

    1 FOO {1.28 USD}

However, the quoted cost was 1.2817 and I think it makes sense to record that.

If instead of buying 1 FOO, I buy 1000 FOO, then those digits of precision are actually relevant:

    Investment     1000 FOO {1.2817 USD}
    Cash          -1281.70 USD

Again, you could argue that I should use whatever cost I paid.  So when I buy 1 FOO when it's quoted at 1.2817 I should just record it as 1.28.  But it seems I'm losing information here... if I want to know the value of FOO on that date, I want to know 1.2817 and not 1.28 (you can argue that I can have both -- a cost of $1.28 for my entry and a separate price entry for $1.2817).

But why should the user do that rounding?  Isn't that what we have programs for?

Another, maybe better example are exchange rates.  They usually have a lot of precision.  I could write a script that looks up an exchange rate and get something like $1.1237 for a Euro and I want to use that value... I don't want to start rounding things manually to say it was $1.12 or $1.13.  I just want to tell beancount that Euro has a certain precision and it should do the right thing for me -- $1.1237 becomes $1.12.

BTW, this is why I, like some others, have have argued in the past that it should be possible to specify both a total cost and a per share cost at the same time.  Then I could say: the cost per share was 1.2817, but actually, since I only bought one, I paid 1.28... or my bank rounds up so I had to pay 1.29.

Something like:

    Investment     1 FOO {1.2817 USD # 1.29 USD}
    Cash          -1.29 USD

    Investment     1000 FOO {1.2817 USD}
    Cash          -1281.70 USD

seems quite elegant to me.

For what it's worth, as I'm writing this, I am wondering if a rounded cost (the actual value I paid) plus a price entry at higher precision makes sense... but it just seems less *elegant* compared to what I propose above.

Basically, I'd like to see the following mechanisms:

1) The ability to specify rounding and precision on a per account/per currency basis so beancount would do the dirty work for me: `1.2817 USD` would become `1.28` or `1.29` internally depending on my configuration.

2) The ability to say something like `1.2817 USD # 1.29 USD` -- in case my rules are sometimes not right and to represent all information.  This looks like a more faithful representation to me compared to what we have at the moment.

Does that make some sense to you?

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Again, I think you don't understand what I wrote. I completely agree with the first half of your last response. The cost is currently  NOT getting rounded.

The only rounding that occurs here is when you don't specify an amount on a posting. It fills in the number. That number gets rounded. If you specify a number, it NEVER gets rounded. It gets stored as is.

About (1): there is NO ROUDING happening right now. Adding rounding  where there isn't will create a lot of unexpected behavior.

About storing the total-cost (2): the amount of complications this would introduce is nuts. I've discussed this many times before. The least inelegant way I can imagine of solving this is by creating some sort of wrapper over a pair of numbers that would do the right thing depending on the operations. But... right now this is causing no real problems; I don't see where using the unrounded number causes problems. If you want 1.29... put 1.29. If you want 1.2817, put that.
